roc.go

```diff
@@ -0,0 +1,136 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true
+// positive rate (TPR) values corresponding to n
+// cutoffs spanning the relative (or receiver)
+// operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+// input y must be sorted, and SortWeightedLabelled can be used
+// in order to sort data y together with classes and weights.
+// For a given cutoff value, observations corresponding to
+// entries in y greater than the cutoff are classified as
+// false, and these assigned class labels are compared to
+// classes to calculate FPR and TPR.
+// weights being nil results in the unweighted case.
+// n of zero results in all possible cutoffs being calculated.
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.IsSorted(sort.Float64Slice(y)) {
+		panic("stat: input must be sorted")
+	}
+
+	var iw, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot less than 2 points on a ROC curve")
+		}
+		iw = (y[len(y)-1] - y[0]) / float64(n-1)
+		if iw <= 0 {
+			panic("stat: y constant or not sorted")
+		}
+		tol = y[0] + iw
+	}
+
+	if n > (len(y) + 1) {
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	newEntry := true
+	var j int
+	var nPositive, nNegative float64
+	for i, u := range classes {
```

Comment:
This algorithm is very confusing to me.

Looking at the first if statement, it appears that newEntry is changed every other iteration. This doesn't make any sense. Upon a deeper look (after saying 'really? how does this possibly work', downloading your code and implementing a few more test cases), I see that this isn't really what's happening.

Point being, there must be a cleaner algorithm to do these implementations. For the moment, this would be helped by
- Giving 'j' a better name, like bin
- It feels like newEntry isn't actually necessary. Maybe you just need a "bin" and "prevBin". I find the name newEntry not very helpful, and given what I know the algorithm needs to do, it's not clear why such a boolean is necessary (which makes it more difficult)
- It's confusing that all of the data is assigned to by "j+1". I think it would be better to have j just start at 1 with a comment "// the initial bin is known to have 0 fpr and 0 tpr
- For the time being, it would be helpful to have the weights == nil check inside the inner to help see the algorithm better.
- Rather than check if n == 0, instead before the loop say that if n == 0, n = len(y)+1